File: pysim8/src/pysim8/sim/mu.py
"""Matrix Unit state: registers, command queue, and command record."""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass

from .memory import MEM_SIZE

__all__ = ["MuRegisters", "MuCommand", "MuQueue", "MU_QUEUE_DEPTH"]

MU_QUEUE_DEPTH = 4


@dataclass(slots=True)
class MuCommand:
    """A fully-resolved MMUL/MMAD command (no register references)."""

    op: int
    fmt: int
    a_col: bool
    b_col: bool
    c_col: bool
    accumulate: bool
    a_addr: int
    b_addr: int
    c_addr: int
    m: int  # rows of A and C
    n: int  # cols of B and C
    k: int  # inner dim


class MuRegisters:
    """MU register file: MA, MB, MC (pointers), MM, MN, MK (dims), MFPSR (flags)."""

    __slots__ = ("ma", "mb", "mc", "mm", "mn", "mk", "mfpsr")

    def __init__(self) -> None:
        self.ma: int = 0
        self.mb: int = 0
        self.mc: int = 0
        self.mm: int = 0
        self.mn: int = 0
        self.mk: int = 0
        self.mfpsr: int = 0

    _PTR_ATTRS = ("ma", "mb", "mc")
    _REG_ATTRS = ("ma", "mb", "mc", "mm", "mn", "mk")

    def read_ptr(self, code: int) -> int:
        """Read pointer register by code (0=MA, 1=MB, 2=MC)."""
        if code >= len(self._PTR_ATTRS):
            raise ValueError(f"Invalid MU pointer code: {code}")
        return getattr(self, self._PTR_ATTRS[code])

    def write_reg(self, code: int, val: int) -> None:
        """Write MU register by code (0=MA..5=MK)."""
        if code >= len(self._REG_ATTRS):
            raise ValueError(f"Invalid MU register code: {code}")
        setattr(self, self._REG_ATTRS[code], val & 0xFFFF)

    def inc_ptr(self, code: int, amount: int) -> None:
        """Increment pointer register by amount (mod 64K)."""
        cur = self.read_ptr(code)
        self.write_reg(code, (cur + amount) % MEM_SIZE)

    def reset(self) -> None:
        """Reset all MU registers to zero."""
        self.ma = self.mb = self.mc = 0
        self.mm = self.mn = self.mk = 0
        self.mfpsr = 0

    def __repr__(self) -> str:
        return (
            f"MU(MA=0x{self.ma:04X} MB=0x{self.mb:04X} MC=0x{self.mc:04X} "
            f"MM={self.mm} MN={self.mn} MK={self.mk} MFPSR=0x{self.mfpsr:02X})"
        )


class MuQueue:
    """Fixed-depth FIFO command queue for MU."""

    __slots__ = ("_q", "_fault")

    def __init__(self) -> None:
        self._q: deque[MuCommand] = deque(maxlen=MU_QUEUE_DEPTH)
        self._fault: int = 0

    @property
    def is_empty(self) -> bool:
        return len(self._q) == 0

    @property
    def is_full(self) -> bool:
        return len(self._q) >= MU_QUEUE_DEPTH

    @property
    def fault(self) -> int:
        return self._fault

    @fault.setter
    def fault(self, code: int) -> None:
        self._fault = code

    def peek(self) -> MuCommand:
        return self._q[0]

    def enqueue(self, cmd: MuCommand) -> None:
        if self.is_full:
            raise RuntimeError("MU queue full")
        self._q.append(cmd)

    def dequeue(self) -> MuCommand:
        return self._q.popleft()

    def flush(self) -> None:
        self._q.clear()

    def reset(self) -> None:
        self._q.clear()
        self._fault = 0

    def __len__(self) -> int:
        return len(self._q)

File: web/lib/mu.js
/**
 * Matrix Unit state: registers, command queue.
 * Pure logic -- no DOM dependencies.
 */

export const MU_QUEUE_DEPTH = 4;

const MEM_SIZE = 65536;

const _PTR_NAMES = ["ma", "mb", "mc"];
const _REG_NAMES = ["ma", "mb", "mc", "mm", "mn", "mk"];

export class MuCommand {
    constructor(op, fmt, aCol, bCol, cCol, accumulate, aAddr, bAddr, cAddr, m, n, k) {
        this.op = op;
        this.fmt = fmt;
        this.aCol = aCol;
        this.bCol = bCol;
        this.cCol = cCol;
        this.accumulate = accumulate;
        this.aAddr = aAddr;
        this.bAddr = bAddr;
        this.cAddr = cAddr;
        this.m = m;
        this.n = n;
        this.k = k;
    }
}

export class MuRegisters {
    constructor() {
        this.ma = 0;
        this.mb = 0;
        this.mc = 0;
        this.mm = 0;
        this.mn = 0;
        this.mk = 0;
        this.mfpsr = 0;
    }

    readPtr(code) {
        return this[_PTR_NAMES[code]];
    }

    writeReg(code, val) {
        this[_REG_NAMES[code]] = val & 0xffff;
    }

    incPtr(code, amount) {
        this.writeReg(code, (this.readPtr(code) + amount) % MEM_SIZE);
    }

    reset() {
        this.ma = this.mb = this.mc = 0;
        this.mm = this.mn = this.mk = 0;
        this.mfpsr = 0;
    }
}

export class MuQueue {
    constructor() {
        this._q = [];
        this.fault = 0;
    }

    get isEmpty() {
        return this._q.length === 0;
    }

    get isFull() {
        return this._q.length >= MU_QUEUE_DEPTH;
    }

    peek() {
        return this._q[0];
    }

    enqueue(cmd) {
        if (this.isFull) throw new Error("MU queue full");
        this._q.push(cmd);
    }

    dequeue() {
        return this._q.shift();
    }

    flush() {
        this._q = [];
    }

    reset() {
        this._q = [];
        this.fault = 0;
    }
}
